component/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::sync::oneshot::{
+    channel as oneshotChannel, Receiver as oneshotReceiver, Sender as oneshotSender,
+};
+
+pub type IrrecoverableError = anyhow::Error;
+type JoinHandle = tokio::task::JoinHandle<()>;
+
+/// A Supervisor is instantiated to supervise a task that should always be running.
+/// A running supervisor will start a component task, and ensure that it is restarted
+/// if it ever stops.
+pub struct Supervisor<M: Manageable> {
+    irrecoverable_signal: Receiver<IrrecoverableError>,
+    cancellation_signal: Option<oneshotSender<()>>,
+    join_handle: Option<JoinHandle>,
+    manageable: M,
+}
+
+/// In order to be Manageable, a user defines the following two functions:
+///
+/// 1. A start function that launches a tokio task, as input it takes:
+/// - an irrecoverable error sender, on which the component sends information to the supervisor about
+/// an irrecoverable event that has occurred
+/// - a cancellation handle, which will be listened to in the task once an irrecoverable message
+/// has been sent, used as an "ack" that the message has been received and so the function can return
+///
+/// 2. A handle_irrecoverable which takes actions on a relaunch due to an irrecoverable error
+/// that happened. It takes the error message that may contain a stack trace and other information
+/// that was sent to the Supervisor via the tx_irrecoverable passed into start.
+#[async_trait]
+pub trait Manageable {
+    // The function that spawns a tokio task
+    async fn start(
+        &self,
+        tx_irrecoverable: Sender<anyhow::Error>,
+        rx_cancellation: oneshotReceiver<()>,
```

Comment:
> dropping the old cancellation_signal implicitly sends cancellation by closing the channel

Ok, now I see this comment, and understand the intention of cancellation. So if I understand correctly, the receiver end will get [TryRecvError](https://docs.rs/tokio/latest/tokio/sync/oneshot/error/enum.TryRecvError.html#) on the channel when this is dropped? This is subtle, so you may want to make an `on_cancelled()` async function in `Manageable` that returns a future that fires when a cancel signal is returned. 

There is a little subtlety here (hence why there is value in the library having that code), in that the rx end of the cancel channel needs to be taken as `&mut`. So a test case for waiting to cancel + other event, then another event happening, then waiting to cancel again, then cancelling, would be welcome.

Reply:
This is indeed very subtle. The purpose of this channel is to avoid a race condition where we send a message on `tx_irrecoverable` and immediately return. In that situation it is a coin flip whether the first branch or the second branch of the `run` method's select statement will get triggered, since the future could return before the message is received, or vice versa. (I have observed this in my testing as well.) So we use the cancellation as an "ack" that the message we sent was received, and then return out of our function. If we don't do this, often times the message we sent will be lost.